data_structures/check_binary_search_tree.py

```diff
@@ -0,0 +1,183 @@
+# check if the binary tree is a binary search tree.
+
+from collections import deque
+
+
+class Node:
+    # create a Node class
+    def __init__(self, data):
+        self.data = data
+        self.left = None
+        self.right = None
+        self.height = 0
+
+    def __str__(self):
+        lines, *_ = self._display_aux()
+        return "\n".join(lines)
+
+    def _display_aux(self):
+        # No child.
+        if self.right is None and self.left is None:
+            if self.height:
+                line = "({}, {})".format(self.data, self.height)
+            else:
+                line = "({})".format(self.data)
+            width = len(line)
+            height = 1
+            middle = width // 2
+            return [line], width, height, middle
+
+        # Only left child.
+        if self.right is None:
+            lines, n, p, x = self.left._display_aux()
+            if self.height > 0:
+                s = "({}, {})".format(self.data, self.height)
+            else:
+                s = "({})".format(self.data)
+            u = len(s)
+            first_line = (x + 1) * " " + (n - x - 1) * "-" + s
+            second_line = x * " " + "/" + (n - x - 1 + u) * " "
+            shifted_lines = [line + u * " " for line in lines]
+            return [first_line, second_line] + shifted_lines, n + u, p + 2, n + u // 2
+
+        # Only right child.
+        if self.left is None:
+            lines, n, p, x = self.right._display_aux()
+            if self.height:
+                s = "({}, {})".format(self.data, self.height)
+            else:
+                s = "({})".format(self.data)
+            u = len(s)
+            first_line = s + x * "-" + (n - x) * " "
+            second_line = (u + x) * " " + "\\" + (n - x - 1) * " "
+            shifted_lines = [u * " " + line for line in lines]
+            return [first_line, second_line] + shifted_lines, n + u, p + 2, u // 2
+
+        # Two children.
+        left, n, p, x = self.left._display_aux()
+        right, m, q, y = self.right._display_aux()
+        if self.height:
+            s = "({}, {})".format(self.data, self.height)
+        else:
+            s = "({})".format(self.data)
+        u = len(s)
+        first_line = (x + 1) * " " + (n - x - 1) * "-" + s + y * "-" + (m - y) * " "
+        second_line = (
+            x * " " + "/" + (n - x - 1 + u + y) * " " + "\\" + (m - y - 1) * " "
+        )
+        if p < q:
+            left += [n * " "] * (q - p)
+        elif q < p:
+            right += [m * " "] * (p - q)
+        zipped_lines = zip(left, right)
+        lines = [first_line, second_line] + [a + u * " " + b for a, b in zipped_lines]
+        return lines, n + m + u, max(p, q) + 2, n + u // 2
+
+
+class Binary_tree:
+    def __init__(self):
+        self.root = None
+
+
+def add_node(sub_root, n1=None, n2=None):
```

Comment:
Why are nont of these functions of the BinarySearchTree.

Reply:
Individual functions move as member functions of the BinaryTree class